Tensile/SolutionStructs.py

```diff
@@ -1720,60 +1721,61 @@ class Solution(collections.abc.Mapping):
     for key in defaultSolution:
-      assignParameterWithDefault(self._state, key, config, defaultSolution)
+      assignParameterWithDefault(state, key, config, defaultSolution)
 
-    if 'ISA' not in self._state:
+    if 'ISA' not in state:
       if 'ISA' in config:
-        self._state['ISA'] = config['ISA']
+        state['ISA'] = config['ISA']
       elif config['KernelLanguage'] == 'Assembly':
-        self._state['ISA'] = list(globalParameters["CurrentISA"])
+        state['ISA'] = list(globalParameters["CurrentISA"])
       else:
-        self._state['ISA'] = [0,0,0]
+        state['ISA'] = [0,0,0]
 
-    if "CodeObjectVersion" not in self._state:
+    if "CodeObjectVersion" not in state:
       if "CodeObjectVersion" in config:
-        self._state["CodeObjectVersion"] = config["CodeObjectVersion"]
+        state["CodeObjectVersion"] = config["CodeObjectVersion"]
       else:
-        self._state["CodeObjectVersion"] = globalParameters["CodeObjectVersion"]
+        state["CodeObjectVersion"] = globalParameters["CodeObjectVersion"]
 
     # assign parameters without defaults
     for key in config:
-      if key != "ProblemType" and key not in self._state:
-        self._state[key] = config[key]
-    self["Valid"] = True
+      if key != "ProblemType" and key not in state:
+        state[key] = config[key]
+    state["Valid"] = True
     # this could prevent OriginalSolution from re-assigning the parameters, save lots of time
-    if "AssignedProblemIndependentDerivedParameters" not in self._state:
-      self["AssignedProblemIndependentDerivedParameters"] = False
-    if "AssignedDerivedParameters" not in self._state:
-      self["AssignedDerivedParameters"] = False
-
-    if self["ProblemType"].convolution:
-        for (key,value) in self["ProblemType"].convolution.solutionParms.items():
-            self._state[key]=value
-    Solution.assignDerivedParameters(self._state)
+    if "AssignedProblemIndependentDerivedParameters" not in state:
+      state["AssignedProblemIndependentDerivedParameters"] = False
+    if "AssignedDerivedParameters" not in state:
+      state["AssignedDerivedParameters"] = False
+
+    if state["ProblemType"].convolution:
+        for (key,value) in state["ProblemType"].convolution.solutionParms.items():
+            state[key]=value
+    Solution.assignDerivedParameters(state)
     self._name = config["CustomKernelName"] if isCustomKernelConfig(config) else None
+    self._state = MappingProxyType(state)
     self.initHelperKernelObjects()
-
+    
   # these keys are copied from ProblemType to internal that may be overridden
   InternalKeys = ["UseSgprForGRO","VectorStore"]
 
 
   ########################################
   # get a list of kernel parameters for this solution
   def getKernels(self):
-    kernel = deepcopy(self)
-    kernel._state.update({"Kernel": True})
+    kernel = self
+    kernel._mutable_state.update({"Kernel": True})
     kernels = []
     kernels.append(kernel)
     return kernels
```

Comment:
This function will now return a reference to the object itself instead of a full copy, which could have side-effects if calling code requires the original Solution object unchanged. This isn't necessarily avoidable, but it would be great if we could consume the original object.